core/runtime/src/main/java/io/quarkus/runtime/graal/DiagnosticPrinter.java

```diff
@@ -70,4 +44,6 @@ public static void printDiagnostics(PrintStream w) {
             }
         }
     }
+
+    protected abstract void printTid(PrintStream w, Thread thread);
```

Comment:
This solution is likely far more complex than necessary.  Before we go to this extreme, could you try wrapping the POSIX-specific code in a call to `com.oracle.svm.core.os.IsDefined#isLinux`/`#isDarwin`?  These values are constant folded and thus allow conditional compilation.

Reply:
You might need to move the platform-specific code to methods.  I've had bugs where it didn't work otherwise, though the GraalVM people couldn't explain why.

Reply:
You could probably have still used the `OS.getCurrent() != OS.WINDOWS` strategy if you put that conditional into a method with the `@Fold` annotation.  However the problem is that there is no artifact containing this annotation, so there's really no way to add it in real code!

Reply:
I just tested `IsDefined` and it does solve the issue in a much simpler way. Thanks for the suggestion @dmlloyd!

There's only one drawback to it: we can't use `IsDefined` to include some Windows-specific code since there's no `IsDefined.isWindows()`, unless we can assume that any OS that is not Linux or Darwin is a Windows OS (which does not sound reasonnable to me).

If we want to retrieve and print the Windows native thread ID, then the complex version might be the only way to do it. If not, the `IsDefined` is definitely the best one.

WDYT?

Reply:
I think that for the time being it'd be OK to assume that `windows == ! linux && ! macos`.  It can always be revisited later, and/or an issue can be opened upstream to request an `IsWindows` method.

One option could be to introduce a utility method somewhere, marked `@AlwaysInline`, which does this logic so that if/when there's a better way to identify Windows it can be updated in just one place.  But I wouldn't worry too much about it otherwise.